src/System.Management.Automation/help/ManpageHelpProvider.cs

```diff
@@ -0,0 +1,160 @@
+#if UNIX
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Globalization;
+using System.IO;
+using System.Reflection;
+using System.Xml;
+
+using System.Diagnostics;
+
+#nullable enable
+
+namespace System.Management.Automation
+{
+    /// <summary>
+    /// Class ProviderHelpProvider implement the help provider for manpages.
+    /// </summary>
+    internal class ManpageHelpProvider : HelpProviderWithCache
+    {
+        /// <summary>
+        /// Constructor for HelpProvider.
+        /// </summary>
+        internal ManpageHelpProvider(HelpSystem helpSystem) : base(helpSystem)
+        {
+            _sessionState = helpSystem.ExecutionContext.SessionState;
+        }
+
+        private readonly SessionState _sessionState;
+
+        #region Common Properties
+
+        /// <summary>
+        /// Name of this help provider.
+        /// </summary>
+        /// <value>Name of this help provider.</value>
+        internal override string Name
+        {
+            get
+            {
+                return "Manpage Help Provider";
+            }
+        }
+
+        /// <summary>
+        /// Help category of this provider.
+        /// </summary>
+        /// <value>Help category of this provider</value>
+        internal override HelpCategory HelpCategory
+        {
+            get
+            {
+                return HelpCategory.Manpage;
+            }
+        }
+
+        #endregion
+
+        #region Help Provider Interface
+
+        /// <summary>
+        /// Do exact match help based on the target.
+        /// </summary>
+        /// <param name="helpRequest">Help request object.</param>
+        internal override IEnumerable<HelpInfo> ExactMatchHelp(HelpRequest helpRequest)
+        {
+            foreach (ManpageInfo manpage in ManpageSearch(helpRequest.Target))
+            {
+                yield return ManpageHelpInfo.GetHelpInfo(manpage);
+            }
+        }
+
+        /// <summary>
+        /// Return manpage info that sutistfies a given target pattern one by one.
+        /// </summary>
+        /// <param name="pattern">Manpage search eacrh patter.</param>
+        public static IEnumerable<ManpageInfo> ManpageSearch(string pattern)
```

Comment:
Since all data are accumulated in Collection<ManpageInfo> it makes no sense to return IEnumerable<ManpageInfo>. If we return  Collection<ManpageInfo> we can simplify a code on above level.
And this should be internal.

Reply:
Changed